src/utils/path-security.ts

```
@@ -0,0 +1,43 @@
+
+    if (!isAllowed) {
+        throw new Error(`Invalid path: must start with one of [${allowedRoots.join(', ')}]`);
+    }
+
+    // Basic character validation (Unreal strictness)
+    // Blocks: < > : " | ? * (Windows reserved) and control characters
+    // allowing spaces, dots, underscores, dashes, slashes
+    // Note: Unreal allows spaces in some contexts but it's often safer to restrict them if strict mode is desired.
+    // For now, we block the definitely invalid ones.
+    // eslint-disable-next-line no-control-regex
+    const invalidChars = /[<>:"|?*\x00-\x1f]/;
```

Biome (2.1.2)

[error] Unexpected control character in a regular expression.

Control characters are unusual and potentially incorrect inputs, so they are disallowed.

(lint/suspicious/noControlCharactersInRegex)